vehicle-python/src/vehicle_lang/compile/tensorflow/_builtins.py

```diff
@@ -0,0 +1,152 @@
+from dataclasses import dataclass
+from fractions import Fraction
+from typing import Any, Sequence, cast
+
+import tensorflow as tf
+from typing_extensions import override
+
+from .._abc import ABCBuiltins
+from .._ast import _nodes
+
+################################################################################
+### Type-safe TensorFlow wrappers
+################################################################################
+
+
+def _tf_constant(*args: Any, **kwargs: Any) -> tf.Tensor:
+    """Type-safe wrapper for tf.constant that casts complex return type to tf.Tensor."""
+    return cast(tf.Tensor, tf.constant(*args, **kwargs))
+
+
+################################################################################
+### Interpretations of Vehicle builtins in Tensorflow
+################################################################################
+
+
+@dataclass(frozen=True)
+class TensorFlowBuiltins(
+    ABCBuiltins[
+        int,
+        float,
+        tf.Tensor,
+    ]
+):
+    dtype_index: tf.DType = tf.uint32
+    dtype_rat: tf.DType = tf.float32
+
+    @override
+    def RatTensor(self, value: _nodes.Tensor) -> tf.Tensor:
+        match value.value:
+            case Fraction():
+                # Single value - expand to tensor shape
+                float_value = float(value.value)
+                return _tf_constant(
+                    value=float_value, dtype=self.dtype_rat, shape=value.shape
+                )
+            case _:
+                # Sequence of values
+                return _tf_constant(
+                    value=tuple(float(val) for val in value.value),
+                    dtype=self.dtype_rat,
+                    shape=value.shape,
+                )
+
+    @override
+    def NegRatTensor(self, x: tf.Tensor) -> tf.Tensor:
+        return tf.negative(x)
+
+    @override
+    def AddRatTensor(self, x: tf.Tensor, y: tf.Tensor) -> tf.Tensor:
+        return tf.add(x, y)
+
+    @override
+    def SubRatTensor(self, x: tf.Tensor, y: tf.Tensor) -> tf.Tensor:
+        return tf.subtract(x, y)
+
+    @override
+    def MulRatTensor(self, x: tf.Tensor, y: tf.Tensor) -> tf.Tensor:
+        return tf.multiply(x, y)
+
+    @override
+    def DivRatTensor(self, x: tf.Tensor, y: tf.Tensor) -> tf.Tensor:
+        return tf.divide(x, y)
+
+    @override
+    def MinRatTensor(self, x: tf.Tensor, y: tf.Tensor) -> tf.Tensor:
+        return tf.minimum(x, y)
+
+    @override
+    def MaxRatTensor(self, x: tf.Tensor, y: tf.Tensor) -> tf.Tensor:
+        return tf.maximum(x, y)
+
+    @override
+    def ReduceAddRatTensor(
+        self, e: float, xs: tf.Tensor | Sequence[tf.Tensor]
+    ) -> tf.Tensor:
+        # e is the identity element (0 for addition), xs is the samples to reduce
```

Comment:
This implementation looks weird to me. Shouldn't the definition of this whole function simply be `tf.add(tf.reduce_sum(xs), e)`? Same for all the reduction operations.